### PR TITLE
Add Exact Cover

### PR DESCRIPTION
**Pull request by hhtong**
Additional work that could be done:
- adding the graph representation of the problem (see [here](https://en.wikipedia.org/wiki/Exact_cover#Graph_representation))
- finding the smallest exact cover (least number of subsets)

### CODE REVIEW

File: dwave_networkx/algorithms/exact_cover.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                for j in range(i-1, -1, -1):
+                    if element in subsets[j]:
+                        bqm.add_interaction(i, j, 2)
+
+    bqm.offset = offset
+    
+    response = sampler.sample(bqm)
+
+    # Get lowest energy sample
+    sample = next(iter(response))
```
Comment (arcondello) [inline dwave_networkx/algorithms/exact_cover.py:163]: this is an old syntax. You can just do `response.first.sample`

File: dwave_networkx/algorithms/exact_cover.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+    result = []
+    for subset_index in sample:
+        if sample[subset_index] > 0:
+            result.append(subsets[subset_index])
+
+    # Validate that result is actually an exact cover
```
Comment (arcondello) [inline dwave_networkx/algorithms/exact_cover.py:170]: Most of the other dnx functions return the solution whether it is correct or not, but document it.

File: dwave_networkx/algorithms/exact_cover.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    >>> dnx.is_exact_cover(problem_set, subsets)
+    False
+
+    """
+    # Verify that problem_set is valid
+    orig_len = len(problem_set)
+    problem_set = set(problem_set)
+    if len(problem_set) != orig_len:
+        raise ValueError("problem_set should not contain any duplicates")
+
```
Comment (randomir) [inline dwave_networkx/algorithms/exact_cover.py:62]: Should you check the subsets are valid as well?

File: dwave_networkx/algorithms/exact_cover.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    if len(problem_set) != orig_len:
+        raise ValueError("problem_set should not contain any duplicates")
+    
+    # Verify that every element in problem_set is contained exactly once in subsets
+    cover = [x for subset in subsets for x in subset]
+
+    if len(cover) != orig_len:
+        return False
+
+    return all(x in cover for x in problem_set)
```
Comment (randomir) [inline dwave_networkx/algorithms/exact_cover.py:69]: Shorter, and maybe a bit faster:
```suggestion
    return problem_set == set.union(*subsets) and orig_len == sum(map(len, subsets))
```

**Review by randomir — APPROVED** [review]
LGTM! Just a few minor suggestions.

**Pull request closed without merging**